### DISCUSSION TITLE
PSRAM boot error messages on ESP32-S3

### DISCUSSION THREAD

**Discussion by smhodge42** [micropython/micropython#16306] [ESP32]
I loaded the generic Micropython v1.23 firmware (non-octal) onto a brand new Espressif ESP32-S3-DevKitC-v1.1 board, variant "N8R8" that has (supposedly) 8MB of PSRAM. Using rshell I opened a REPL session, pushed the reset button on the board and got the following error messages (in red font):

E (287) quad_psram: PSRAM ID read error: 0x00ffffff, PSRAM chip not found or not supported, or wrong PSRAM line mode
E (288) esp_psram: PSRAM enabled but initialization failed. Bailing out.

It is my understanding that Micropython should automatically handle the presence or absence of PSRAM and configure it for use, but it obviously isn't doing that. Why not? Is there something else that needs to be done? 

For a couple of years now, I have had the same (more or less, IIRC) error messages on the same DevKit but one without any PSRAM (and with earlier Micropython firmware versions as well) but I've just ignored them as the board works just fine otherwise. Why doesn't Micropython gracefully accept that there is no PSRAM present in that case and proceed without any scary looking red error messages?

I've searched discussions but haven't found anything that talks about this that is understandable by someone like me who doesn't know what goes on under the hood with Micropython (let alone gives a clue about how to solve it) .

I'd sure appreciate some help on the matter, because now I need to use the PSRAM. Thanks. The REPL session is below.

Using buffer-size of 32
Connecting to COM22 (buffer-size 32)...
Trying to connect to REPL  connected
Retrieving sysname ... esp32
Testing if ubinascii.unhexlify exists ... Y
Retrieving root directories ... /boot.py/
Setting time ... Nov 26, 2024 08:20:29
Evaluating board_name ... pyboard
Retrieving time epoch ... Jan 01, 2000
Welcome to rshell. Use the exit command to exit rshell.
D:\steve\Projects\uPython> repl
Entering REPL. Use Control-X to exit.
>
MicroPython v1.23.0 on 2024-06-02; Generic ESP32S3 module with ESP32S3
Type "help()" for more information.
>>>
>>>
>>> ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0x2a (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3810,len:0xf8c
load:0x403c9700,len:0xb3c
load:0x403cc700,len:0x2dd4
entry 0x403c989c
E (287) quad_psram: PSRAM ID read error: 0x00ffffff, PSRAM chip not found or not supported, or wrong PSRAM line mode
E (288) esp_psram: PSRAM enabled but initialization failed. Bailing out.
MicroPython v1.23.0 on 2024-06-02; Generic ESP32S3 module with ESP32S3
Type "help()" for more information.
>>>

**Comment by shariltumin** [comment]
According to https://docs.espressif.com/projects/esp-idf/en/stable/esp32s3/hw-reference/esp32s3/user-guide-devkitc-1.html


ESP32-S3-WROOM-1-N8R8: Flash 8 MB QD, PSRAM 8 MB OT 


So I think you should download and try the "Firmware (Support for Octal-SPIRAM)".

**Comment by smhodge42** [comment]
Thank you, that solved it, at least on the new DevKit. I misread the specs and thought it was not octal.

However, using the OT version does not solve it on my earlier boards, the ones I referred to as having no PSRAM. I still get the error messages with them. Curiously, on close inspection of the ESP chip on those boards it says “P2N8”, and the “P2” part is NOT anything listed on the Espressif link below even though they are – supposedly – genuine Espressif parts.

Do you have any idea what the “P2” means? I was unable to find it with a web search.  Perhaps I have some sort of oddball version that gives Micropython trouble.  I’ll poke around and see if I have a simple “N8” (no P2 part) variant somewhere.

**Reply by shariltumin** [reply]
I do not know what P2N8 means. I have a board that has the P2N8 and nothing else. While flashing I saw "Auto-detected Flash size: 16MB". I am not sure about the RAM size.

Try, https://micropython.org/resources/firmware/ESP32_GENERIC_S3-20241030-v1.25.0-preview.6.g548babf8a.bin

```bash
MicroPython v1.25.0-preview.6.g548babf8a on 2024-10-30; Generic ESP32S3 module with ESP32S3
Type "help()" for more information.
>>> import micropython as mp
>>> mp.mem_info(1)
stack: 720 out of 15360
GC: total: 64000, used: 1888, free: 62112, max new split: 155648
 No. of 1-blocks: 29, 2-blocks: 7, max blk sz: 18, max free sz: 3847
GC memory layout; from 3fcb2eb0:
00000000: h=hLhhhBMhhDhhhh=hhh=================hh=======h=======h=hh======
00000400: =========Bh=h=hhhBhhhhh==B..hhh=.......h=...h==.................
00000800: ......h=======h==========.......................................
       (59 lines all free)
0000f800: ................................
>>> import gc
>>> gc.mem_free()
217456
>>> a = bytearray(600000)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
MemoryError: memory allocation failed, allocating 600000 bytes
>>> a = bytearray(200000)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
MemoryError: memory allocation failed, allocating 200000 bytes
>>> a = bytearray(100000)
>>> gc.mem_free()
118688
>>> mp.mem_info()
stack: 704 out of 15360
GC: total: 164096, used: 101888, free: 62208, max new split: 56320
 No. of 1-blocks: 33, 2-blocks: 9, max blk sz: 6250, max free sz: 3847
>>> 
```

**Comment by ekspla** [comment]
There was [an answer in ESPRESSIF's forum](https://esp32.com/viewtopic.php?t=36381) as follows: 

> It's likely a manufacturing code rather than part of the model name.

**Comment by smhodge42** [comment]
I was guessing the same, a manufacturing code. My N8R8 board, the one that with your help I got to work with the octal PSRAM, actually has “MO” in front of it.

At least now I can get MP to recognize PSRAM when PSRAM is present. The fact that it still doesn’t recognize when it’s not present and issues an error message is annoying and disconcerting but nonetheless can be ignored since the board otherwise works ok.

**Reply by projectgus** [reply]
> At least now I can get MP to recognize PSRAM when PSRAM is present. The fact that it still doesn’t recognize when it’s not present and issues an error message is annoying and disconcerting but nonetheless can be ignored since the board otherwise works ok.

Glad you got everything working OK. FWIW there is an open issue about the bootloader log message for missing PSRAM, with some discussion of the different trade-offs: https://github.com/micropython/micropython/issues/15537
